server/src/main/java/org/opensearch/transport/InboundHandler.java

```diff
@@ -184,79 +192,83 @@ private <T extends TransportRequest> void handleRequest(TcpChannel channel, Head
-            } catch (Exception e) {
-                if (Version.CURRENT.isCompatible(header.getVersion())) {
-                    sendErrorResponse(action, transportChannel, e);
-                } else {
-                    logger.warn(
-                        new ParameterizedMessage(
-                            "could not send error response to handshake received on [{}] using wire format version [{}], closing channel",
-                            channel,
-                            header.getVersion()
-                        ),
-                        e
-                    );
-                    channel.close();
-                }
-            }
-        } else {
-            final TransportChannel transportChannel = new TcpTransportChannel(
-                outboundHandler,
-                channel,
-                action,
-                requestId,
-                version,
-                header.getFeatures(),
-                header.isCompressed(),
-                header.isHandshake(),
-                message.takeBreakerReleaseControl()
-            );
-            try {
+        Span span = tracer.startSpan(SpanBuilder.from(action, channel));
+        channel = TraceableTcpChannel.create(channel, span, tracer);
+        try (SpanScope spanScope = tracer.withSpanInScope(span)) {
+            if (header.isHandshake()) {
                 messageListener.onRequestReceived(requestId, action);
-                if (message.isShortCircuit()) {
-                    sendErrorResponse(action, transportChannel, message.getException());
-                } else {
-                    final StreamInput stream = namedWriteableStream(message.openOrGetStreamInput());
-                    assertRemoteVersion(stream, header.getVersion());
-                    final RequestHandlerRegistry<T> reg = requestHandlers.getHandler(action);
-                    assert reg != null;
-
-                    final T request = newRequest(requestId, action, stream, reg);
-                    request.remoteAddress(new TransportAddress(channel.getRemoteAddress()));
-                    checkStreamIsFullyConsumed(requestId, action, stream);
-
-                    final String executor = reg.getExecutor();
-                    if (ThreadPool.Names.SAME.equals(executor)) {
-                        try {
-                            reg.processMessageReceived(request, transportChannel);
-                        } catch (Exception e) {
-                            sendErrorResponse(reg.getAction(), transportChannel, e);
-                        }
+                // Cannot short circuit handshakes
+                assert message.isShortCircuit() == false;
+                final StreamInput stream = namedWriteableStream(message.openOrGetStreamInput());
+                assertRemoteVersion(stream, header.getVersion());
+                final TransportChannel transportChannel = new TcpTransportChannel(
```

Comment:
Initially I tried with TransportChannel only but seems it's not getting closed on node::close call and keep the spans ended. Let me revisit it again and will get back to you. Thanks!

Reply:
Intuitively, we may not get notification on channel closure but sendResponse / sendError callbacks should fail trying to write into closed channel.

Reply:
@reta, Yes, ideally it should be. But in the following scenarios if lifecycle::started==false then we are just returning and not executing any onSuccess/onFailure operations hence not calling the sendResponse/sendError of TransportChannel. It's fine because nothing can be done if case node is stopped but we especially for tracing need some handler to be executed to clear up the resources in this case. We may follow the approach similar to TransportResponseHandler and provide the handleRejection method also may throw the exception like being done on couple of other places. 

https://github.com/opensearch-project/OpenSearch/blob/cbff21d7ea58d3185b969df43f48de10e32180e2/server/src/main/java/org/opensearch/cluster/service/MasterService.java#L977

https://github.com/opensearch-project/OpenSearch/blob/cbff21d7ea58d3185b969df43f48de10e32180e2/server/src/main/java/org/opensearch/cluster/service/ClusterApplierService.java#L375

Your thoughts on this please. Thanks!

Reply:
Hm ... this is very concerning, I think what we could do in this case, the `TraceableTransportChannel` (the one we need to create) could add itself onCloseListerner to underlying TcpChannel, and than just track if send t sendResponse / sendError  were called or channel was closed, whatever comes first. What do you think?

Reply:
Exactly the same I was thinking. But just wanted to understand the Lifecycle of a TcpChannel, Will it be reused or not and we shouldn't be overloading with a lot of onClose listeners.

Reply:
I am a little skeptical here while adding the onCloseListener for each request. It will overload the TcpChannel, at max we can keep some logic which can clear up all the open TransportChannels that too might be resource consuming.

Reply:
@reta, your thoughts on this?

Reply:
@Gaganjuneja I was looking into the effect of  ``if (!lifecycle.started()) {` checks in MasterService & co and I don't see how it impacts the channel callbacks: even if lifecycle is not started, the empty response is being sent (like in `ShardStartedTransportHandler` fe):

```
 channel.sendResponse(TransportResponse.Empty.INSTANCE);
```

> But in the following scenarios if lifecycle::started==false then we are just returning and not executing any onSuccess/onFailure operations hence not calling the sendResponse/sendError of TransportChannel.

AFAIK the `ClusterApplyListener` does not trigger the writes to transport channel directly (at least I haven't found any) do you have a test case or hint on where exactly the onSuccess/onFailure operations are triggering the transport channel write ?

Reply:
Coordinator registers the PublicationTransportHandler

https://github.com/opensearch-project/OpenSearch/blob/797def6e50389d88634b6ca614ceaba407ee983f/server/src/main/java/org/opensearch/cluster/coordination/Coordinator.java#L257




PublicationTransportHandler registers the internal:cluster/coordination/commit_state

https://github.com/opensearch-project/OpenSearch/blob/797def6e50389d88634b6ca614ceaba407ee983f/server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java#L126

transportCommitCallback

https://github.com/opensearch-project/OpenSearch/blob/797def6e50389d88634b6ca614ceaba407ee983f/server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java#L130

Reply:
Thanks @Gaganjuneja , I think at this moment we don't have any other choice but add a `onCloseListener` guard to `TcpChannel`. We could probably change the `ClusterApplyListener` to handle the lifecycle check but it is mapped over `ActionListener` that has no mechanism to handle that.

Reply:
Yes @reta, I anticipate that we may need to add a new method in the ActionListener::handleRejection as and when we go deep with the instrumentation.